src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,668 @@
+
+        > x * y * z == x * y + x * z + y * z
+
+    * 'connect' has the same effect as 'overlay' on vertices of one part:
+
+        > (leftVertex x)  * (leftVertex y)  == (leftVertex x)  + (leftVertex y)
+        > (rightVertex x) * (rightVertex y) == (rightVertex x) + (rightVertex y)
+
+The following useful theorems can be proved from the above set of axioms.
+
+    * 'overlay' has 'empty'
+    as the identity and is idempotent:
+
+        > x + empty == x
+        > empty + x == x
+        >     x + x == x
+
+    * Absorption and saturation of 'connect':
+
+        > x * y + x + y == x * y
+        >     x * x * x == x * x
+
+When specifying the time and memory complexity of graph algorithms, /n/ and /m/
+will denote the number of vertices and edges in the graph, respectively. In
+addition, /l/ and /r/ will denote the number of vertices in the left and in the
+right part of graph, respectively.
+
+-}
+data AdjacencyMap a b = BAM {
+    -- | The /adjacency map/ of the left part of the graph: each vertex is
+    -- associated with a set of its neighbours. Complexity: /O(1)/ time and
+    -- memory.
+    --
+    -- @
+    -- leftAdjacencyMap 'empty'                    == Map.'Map.empty'
+    -- leftAdjacencyMap ('leftVertex' 1)           == Map.'Map.singleton' 1 Set.'Set.empty'
+    -- leftAdjacencyMap ('rightVertex' 1)          == Map.'Map.empty'
+    -- leftAdjacencyMap ('edge' 1 1)               == Map.'Map.singleton' 1 (Set.'Set.singleton' 1)
+    -- leftAdjacencyMap ('edge' 1 "a")             == Map.'Map.singleton' 1 (Set.'Set.singleton' "a")
+    -- leftAdjacencyMap ('edges' [(1, 1), (1, 2)]) == Map.'Map.singleton' 1 (Set.'Set.fromAscList' [1, 2])
+    -- @
+    leftAdjacencyMap :: Map.Map a (Set.Set b),
+
+    -- | The inverse map for 'leftAdjacencyMap'. Complexity: /O(1)/ time and memory.
+    --
+    -- @
+    -- rightAdjacencyMap 'empty'                    == Map.'Map.empty'
+    -- rightAdjacencyMap ('leftVertex' 1)           == Map.'Map.empty'
+    -- rightAdjacencyMap ('rightVertex' 1)          == Map.'Map.singleton' 2 Set.'Set.empty'
+    -- rightAdjacencyMap ('edge' 1 1)               == Map.'Map.singleton' 1 (Set.'Set.singleton' "a")
+    -- rightAdjacencyMap ('edge' 1 "a")             == Map.'Map.singleton' "a" (Set.'Set.singleton' 1)
+    -- rightAdjacencyMap ('edges' [(1, 1), (1, 2)]) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+    -- @
+    rightAdjacencyMap :: Map.Map b (Set.Set a)
+} deriving Generic
+
+-- | __Note:__ this does not satisfy the usual ring laws; see 'AdjacencyMap'
+-- for more details.
+instance (Ord a, Ord b, Num b) => Num (AdjacencyMap a b) where
+    fromInteger = rightVertex . fromInteger
+    (+)         = overlay
+    (*)         = connect
+    signum      = const empty
+    abs         = id
+    negate      = id
+
+instance (Ord a, Ord b) => Eq (AdjacencyMap a b) where
+    (BAM lr1 rl1) == (BAM lr2 rl2) = (lr1 == lr2) && (Map.keysSet rl1 == Map.keysSet rl2)
```

Comment:
This is a tricky definition. It would be nice to test that it actually produces the same result as the trivial implementation based on comparing both maps.